src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+
+            uint256 oldTipHash = this.peerTipsByPeerId.TryGet(networkPeerId);
+
+            var listOfPeersClaimingThisHeader = this.peerIdsByTipHash.TryGet(newTip);
+            if (listOfPeersClaimingThisHeader == null)
+            {
+                listOfPeersClaimingThisHeader = new HashSet<int>();
+                this.peerIdsByTipHash.Add(newTip, listOfPeersClaimingThisHeader);
+            }
+
+            listOfPeersClaimingThisHeader.Add(networkPeerId);
+            this.peerTipsByPeerId.AddOrReplace(networkPeerId, newTip);
+
+            if (oldTipHash != null)
+            {
+                ChainedHeader oldTip = this.chainedHeadersByHash.TryGet(oldTipHash);
+                this.RemovePeerClaim(networkPeerId, oldTip);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Find the headers that are not part of the tree and try to connect them to an existing chain 
+        /// by creating a new <see cref="ChainedHeader"/> type and linking it to its previous header.
+        /// </summary>
+        /// <remarks>
+        /// A new header will perform partial validation.
+        /// This will take in to account the MaxReorg protection rule, chains that are beyond the max reorg flag will be abandoned.
+        /// This will append to the ChainedHeader.Next of the previous header.
+        /// </remarks>
+        /// <param name="headers">The new headers that should be connected to a chain.</param>
+        /// <returns>A list of newly created <see cref="ChainedHeader"/> or <c>null</c> if no new headers were found.</returns>
+        private List<ChainedHeader> CreateNewHeaders(List<BlockHeader> headers)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(headers), nameof(headers.Count), headers.Count);
+
+            if (!this.FindFirstNewHeader(headers, out int newHeaderPosition))
+            {
+                this.logger.LogTrace("(-)[NO_NEW_HEADERS_FOUND]:null");
+                return null;
+            }
+
+            ChainedHeader previousChainedHeader = this.chainedHeadersByHash.TryGet(headers[newHeaderPosition].HashPrevBlock);
+            if (previousChainedHeader == null)
+            {
+                this.logger.LogTrace("(-)[PREVIOUS_HEADER_NOT_FOUND]: Previous hash `{0}` of block hash `{1}` was not found.", headers[newHeaderPosition].GetHash(), headers[newHeaderPosition].HashPrevBlock);
+                throw new ConnectHeaderException();
+            }
+
+            var newChainedHeaders = new List<ChainedHeader>();
+
+            ChainedHeader newChainedHeader = this.CreateAndValidateNewChainedHeader(newChainedHeaders, headers[newHeaderPosition], previousChainedHeader);
+            newHeaderPosition++;
+            this.logger.LogTrace("New chained header was added to the tree '{0}'.", newChainedHeader);
+
+            this.CheckMaxReorgRuleViolated(newChainedHeader);
```

Comment:
Here we have a problem did we don't get rid of the chained header that we just added to the tree, so this is memory leak.

Reply:
Probably instead of throwing let it return bool and in case it's true- remove invalid header and throw after that